### DISCUSSION TITLE
[PSD12] Exercise 3.24

### DISCUSSION THREAD

**Discussion by matypist** [sapienzastudentsnetwork/psd2223#184] [Esercizi (A-L)]
**Exercise 3.24** _Accidents are still occurring at the intersection of Academic Avenue
and Bravado Boulevard. The football team is rushing into the intersection the
moment light B turns green. They are colliding with sleep-deprived CS majors
who stagger into the intersection just before light A turns red._ Extend the light controller described during the **last lesson** so that both lights are red for 5 seconds before either light turns green again. Sketch your improved Moore machine state transition diagram, state encodings, state transition table, output table, next state and output equations, and your FSM schematic.

- _**N.B.:** La parte in corsivo era stata tagliata dal professore dal testo della consegna (rispetto a quello originario del libro). Per **last lesson** il professore si riferisce alle **slide [13, 26]** del PowerPoint ("**PSD12.pptx**"); nel caso del libro si tratta invece della **sezione 3.4.1**._

**Comment by fraxb04** [comment]
ragazzi aiuto

**Comment by Elia-Belli** [comment]
## Esercizio Originale
**Testo** : Engineering students are moseying between their dorms and the labs on Academic Ave. They are busy reading about FSMs in their favorite textbook and aren’t looking where they are going. Football players are hustling between the athletic fields and the dining hall on Bravado Boulevard. They are tossing the ball back and forth and aren’t looking where they are going eitherr. Several serious injuries have already occurred at the intersection of these two roads, and the Dean of Students asks Ben Bitdiddle to install a traffic light before there are fatalities.

![traffic street](https://user-images.githubusercontent.com/107944476/214876287-3f156cff-3be8-450f-8779-f7d29a6d4520.png)

### Soluzione
Ben decides to solve the problem with an FSM : he installs two traffic sensors, TA and TB, on Academic Ave. and Bravado Blvd., respectively. Each sensor indicates TRUE if students are present and FALSE if the street is empty. He also installs two traffic lights, LA and LB, to control traffic. Each light receives digital inputs specifying whether it should be green, yellow, or red. Ben provides a clock with a 5-second period. On each clock tick (rising edge), the lights may change based on the traffic sensors. He also provides a reset button so that Physical Plant technicians can put the controller in a known initial state when they turn it on.

![traffic diagram](https://user-images.githubusercontent.com/107944476/214874925-5fc1f67f-783f-4df9-ba92-9e108154ee6c.png)

Queste informazioni dovrebbero essere sufficienti a risolvere il nuovo esercizio, per la soluzione completa della versione originale fare riferimento alle pagine 123-128 del libro di testo.

**Comment by Elia-Belli** [comment]
## Interpretazione
Il problema in cui incorriamo è che non appena il semaforo di una strada diventa verde le auto iniziato a passare l'incrocio, ma ci sono ancora persone a cui non bastano i 5 s di giallo per finire di attraversare. Quindi vorremmo che entrambe le luci rimangano rosse per altri 5 s prima che una delle due diventi di nuovo verde.

## Soluzione
Per fare questo, siccome l'FSM che vogliamo modificare è di Moore, aggiungo 2 nuovi stati, $S1T$ e $S3T$, durante i quali entrambi gli output sono rossi ( $L_A$ e $L_B$ ). 

**Osservazione**: "Perchè aggiungere 2 stati e non 1 se fanno la stessa cosa?", il problema non è l'output, ma decidere lo stato successivo, per farlo dobbiamo basarci sugli input ricevuti dagli stati $S1$ o $S3$ in modo da spegnere il semaforo che era in funzione in quel momento e accendere l'altro: in quei 5 secondi di attesa in più potrebbe essersi ricreato del traffico anche sulla strada che si era liberata e quindi con entrambi gli input positivi non sapremmo quale semaforo rendere verde, indipercui il prossimo stato non dipende da nuovi input ma solo dai precedenti.

![TD](https://user-images.githubusercontent.com/107944476/214940548-5c4c4302-4203-4b4d-8e0e-3a24e97e5099.jpg)

**State Encodings**: $S0: 000;\ S1:001;\ S1T: 010;\ S2: 011;\ S3:100;\ S3T:101;$
**Output Encodings**: Red: $00$; Green: $01$; Yellow: $10$;

| $S_2$ | $S_1$ | $S_0$ | $T_A$ | $T_B$ | $S_2'$ | $S_1'$ | $S_0'$ | $LA_1$ | $LA_0$ | $LB_1$ | $LB_0$ |
| --- | ---  | ---  | ---  | --- | ---  | ---  | ---  | ---  | ---  | --- | --- |
|   0  | 0   |   0  |  0   |  X   |  0   | 0    |    1 |  0   |   1  |   0  | 0    |
|  0   |  0   |  0   |  1   |  X   |  0   |   0  |   0  |   0   |  1   |   0  | 0    |
|  0   |  0   |  1   |   X  |   X  |   0  |   1  |  0   |  1   |  0   |  0   |  0   |
|  0   |  1   |  0   |   X  |   X  |   0  |   1  |  1   |  0   |  0   |  0   |  0   |
|   0  | 1   |   1  |  X   |  0   |  1  | 0    |    0 |  0   |   0  |   0  | 1    |
|  0   |  1  |  1   |  X   |  1   |  0   |   1  |   1  |   0   |  0   |   0  | 1    |
|  1   |  0   |  0   |   X  |   X  |  1   |   0  |  1   |  0   |  0   |  1   |  0   |
|  1  |  0   |  1   |   X  |   X  |   0  |   0  |  0   |  0   |  0   |  0   |  0   |

Tramite delle K-map a 5 variabili potremmo ricavarci le funzioni minime, ma non è richiesto, quindi lascio l'esercizio al lettore.

$S_2' = \bar S_2 S_1 S_0 \bar T_B + S_2 \bar S_1 \bar S_0$
$S_1' = \bar S_2 \bar S_1 S_0 + \bar S_2 S_1 \bar S_0 + \bar S_2 S_1 S_0 T_B$
$S_0' = \bar S_2 \bar S_1 \bar S_0 \bar T_A + \bar S_2 S_1 \bar S_0 + \bar S_2 S_1 S_0 T_B + S_2 \bar S_1 \bar S_0$
$LA_1 = \bar S_2 \bar S_1 S_0$
$LA_0 = \bar S_2 \bar S_1 \bar S_0$
$LB_1 = S_2 \bar S_1 \bar S_0$
$LB_0 = \bar S_2 S_1 S_0$

![TC](https://user-images.githubusercontent.com/107944476/214949523-f2d8ec45-43b6-4539-a785-1b9d36ad954d.jpg)

**Comment by FeddyLix17** [comment]
@Elia-Belli a grandi linee come avresti ricavato le funzioni minime usando k-maps a 5 variabili? 
Ci ho provato sbagliando in entrambi i casi, quindi ho semplificato utilizzando i vari teoremi

- **S<sup>'</sup><sub>2</sub>**  non è ulteriormente semplificabile

- **S<sup>'</sup><sub>1</sub>** = S&#x0304;<sub>2</sub>S&#x0304;<sub>1</sub>S<sub>0</sub> + S&#x0304;<sub>2</sub>S<sub>1</sub>S&#x0304;<sub>0</sub> + S&#x0304;<sub>2</sub>S<sub>1</sub>S<sub>0</sub>T<sub>B</sub>
S&#x0304;<sub>2</sub>S<sub>0</sub> ( S&#x0304;<sub>1</sub> + S<sub>1</sub>T<sub>B</sub> ) + S&#x0304;<sub>2</sub>S<sub>1</sub>S&#x0304;<sub>0</sub>
S&#x0304;<sub>2</sub>S&#x0304;<sub>1</sub>S<sub>0</sub> + S&#x0304;<sub>2</sub>S<sub>0</sub>T<sub>B</sub> + S&#x0304;<sub>2</sub>S<sub>1</sub>S&#x0304;<sub>0</sub>
- **S<sup>'</sup><sub>0</sub>** = S&#x0304;<sub>2</sub>S&#x0304;<sub>1</sub>S&#x0304;<sub>0</sub>T&#x0304;<sub>A</sub> + S&#x0304;<sub>2</sub>S<sub>1</sub>S&#x0304;<sub>0</sub> + S&#x0304;<sub>2</sub>S<sub>1</sub>S<sub>0</sub>T<sub>B</sub> + S<sub>2</sub>S&#x0304;<sub>1</sub>S&#x0304;<sub>0</sub>
S&#x0304;<sub>2</sub>S&#x0304;<sub>0</sub> ( S<sub>1</sub> + S&#x0304;<sub>1</sub>T&#x0304;<sub>A</sub> ) + S&#x0304;<sub>2</sub>S<sub>1</sub>S<sub>0</sub>T<sub>B</sub> + S<sub>2</sub>S&#x0304;<sub>1</sub>S&#x0304;<sub>0</sub>
S&#x0304;<sub>2</sub>S&#x0304;<sub>0</sub> ( S<sub>1</sub> + T&#x0304;<sub>A</sub> ) + S&#x0304;<sub>2</sub>S<sub>1</sub>S<sub>0</sub>T<sub>B</sub> + S<sub>2</sub>S&#x0304;<sub>1</sub>S&#x0304;<sub>0</sub>
S&#x0304;<sub>2</sub>S<sub>1</sub>S&#x0304;<sub>0</sub> + S&#x0304;<sub>2</sub>S&#x0304;<sub>0</sub>T&#x0304;<sub>A</sub> +  S&#x0304;<sub>2</sub>S<sub>1</sub>S<sub>0</sub>T<sub>B</sub> + S<sub>2</sub>S&#x0304;<sub>1</sub>S&#x0304;<sub>0</sub>
S&#x0304;<sub>2</sub>S<sub>1</sub> ( S&#x0304;<sub>0</sub> + S<sub>0</sub>T<sub>B</sub> ) + S&#x0304;<sub>2</sub>S&#x0304;<sub>0</sub>T&#x0304;<sub>A</sub> + S<sub>2</sub>S&#x0304;<sub>1</sub>S&#x0304;<sub>0</sub>
S&#x0304;<sub>2</sub>S<sub>1</sub> ( S&#x0304;<sub>0</sub> + T<sub>B</sub> ) + S&#x0304;<sub>2</sub>S&#x0304;<sub>0</sub>T&#x0304;<sub>A</sub> + S<sub>2</sub>S&#x0304;<sub>1</sub>S&#x0304;<sub>0</sub>
S&#x0304;<sub>2</sub>S<sub>1</sub>S&#x0304;<sub>0</sub> + S&#x0304;<sub>2</sub>S<sub>1</sub>T<sub>B</sub> + S&#x0304;<sub>2</sub>S&#x0304;<sub>0</sub>T&#x0304;<sub>A</sub> + S<sub>2</sub>S&#x0304;<sub>1</sub>S&#x0304;<sub>0</sub>


_perdonate il disordine ma questo dovrebbe essere il circuito risultante (molto simile a quello di elia)_

![Circuito psd12 es 3 24 ](https://user-images.githubusercontent.com/66442720/215460294-6fd619dd-17fe-41f1-a0e7-cc65cdfde0f5.jpg)

**Comment by Elia-Belli** [comment]
@FedVlogger17 ti faccio l'esempio per la funzione di $S_2'$.
- Come in una comune k-map ti inizi a riscrivere i mintermini che appaiono nell'**Encoded Transition Table** nella casella corrispondente della k-map: se hai mintermini del tipo $0000X$ vuol dire che devi mettere quel valore in $00000$ e $00001$.
- Riportati tutti i valori della tabella (sia 0 che 1) potrebbero rimanere alcune caselle vuote perchè la FSM ha degli stati irragiungibili inutilizzati, quindi lì mettiamo dei Dont'Care X.
![Immagine 2023-01-30 122946](https://user-images.githubusercontent.com/107944476/215466310-9a698312-b7f2-47fc-9c2e-fa67f91da88a.png)
- A questo punto evidenziamo gli implicanti primi e scriviamo l'equazione semplificata: in questo caso uno da 8 e  uno da 4
$S_2' =S_2 \bar S_0 + S_1 S_0 \bar T_B$
- Ci rendiamo conto quindi che la funzione di $S_2'$ è ulteriormente semplificabile sfruttando i Don't Care.
